File: DSA/Stack/Code/PeekOperation.py
# Peek Operation in Python

class Stack:
    def __init__(self, max_size=100):
        self.arr = [0] * max_size
        self.top = -1
        self.max_size = max_size
    
    def is_empty(self):
        return self.top == -1
    
    def push(self, value):
        self.top += 1
        self.arr[self.top] = value
    
    def peek(self):
        if self.is_empty():
            print("Stack is empty - cannot peek")
            return -1
        print(f"Peeked: {self.arr[self.top]} (at position {self.top})")
        return self.arr[self.top]
    
    def display(self):
        if self.is_empty():
            print("Stack is empty")
            return
        print("Current stack: ", end="")
        for i in range(self.top + 1):
            print(self.arr[i], end=" ")
        print()

# Test
stack = Stack()

print("=== Peek Operation ===\n")

stack.peek()

print("\nPushing elements:")
for i in range(1, 6):
    stack.push(i * 10)

stack.display()

print("\nPeeking without removing:")
for i in range(3):
    stack.peek()

stack.display()

print("\nComplexity Analysis:")
print("Time: O(1)")
print("Space: O(1)")

File: DSA/Stack/Code/PeekOperation.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#define MAX_SIZE 100

// Peek Operation in C
typedef struct {
    int arr[MAX_SIZE];
    int top;
} Stack;

Stack* createStack() {
    Stack* stack = (Stack*)malloc(sizeof(Stack));
    stack->top = -1;
    return stack;
}

bool isEmpty(Stack* stack) {
    return stack->top == -1;
}

void push(Stack* stack, int value) {
    stack->arr[++stack->top] = value;
}

int peek(Stack* stack) {
    if (isEmpty(stack)) {
        printf("Stack is empty - cannot peek\n");
        return -1;
    }
    printf("Peeked: %d (at position %d)\n", stack->arr[stack->top], stack->top);
    return stack->arr[stack->top];
}

void display(Stack* stack) {
    if (isEmpty(stack)) {
        printf("Stack is empty\n");
        return;
    }
    printf("Current stack: ");
    for (int i = 0; i <= stack->top; i++) {
        printf("%d ", stack->arr[i]);
    }
    printf("\n");
}

int main() {
    Stack* stack = createStack();
    
    printf("=== Peek Operation ===\n\n");
    
    peek(stack); // Try to peek empty stack
    
    printf("\nPushing elements:\n");
    for (int i = 1; i <= 5; i++) {
        push(stack, i * 10);
    }
    
    display(stack);
    
    printf("\nPeeking without removing:\n");
    for (int i = 0; i < 3; i++) {
        peek(stack);
    }
    
    display(stack);
    
    printf("\nComplexity Analysis:\n");
    printf("Time: O(1)\n");
    printf("Space: O(1)\n");
    
    free(stack);
    return 0;
}
